man/unigen-hangul.1
.TH UNIGEN-HANGUL 1 "30 July 2023"
.SH NAME
unigen-hangul \- Generate Hangul syllables from a Johab 6/3/1 Unifont hex file
.SH SYNOPSIS
\fBunigen-hangul \-i\fP \fIhangul-base.hex \fP
\fB\-o\fP \fIhangul-syllables.hex\fP
.SH DESCRIPTION
.B unigen-hangul
generates Hangul syllables from an input Unifont .hex file encoded
in Johab 6/3/1 format.  By default, the output is the Unicode Hangul
Syllables range, U+AC00..U+D7A3.  Options allow the user to specify
a starting code point for the output Unifont .hex file, and ranges
in hexadecimal of the starting and ending Hangul Jamo code points:
.RS
.IP \fBRange\fP 15
\fBHangul\fP
.TP
.BR 1100\-115E
Hangul Jamo initial consonants (choseong)
.TP
.BR A960\-A97C
Hangul Jamo Extended-A initial consonants (choseong)
.TP
.BR 1161\-11A7
Hangul Jamo medial vowels and diphthongs (jungseong)
.TP
.BR D7B0\-D7C6
Hangul Jamo Extended-B medial vowels and diphthongs (jungseong)
.TP
.BR 11A8\-11FF
Hangul Jamo final consonants (jongseong).
.TP
.BR D7CB\-D7FB
Hangul Jamo Extended-B final consonants (jongseong).
.RE
.PP
A single code point or 0 to omit can be specified instead of a range.
A starting code point of one position before a valid starting range
for a Hangul jamo series (choseong, jungseong, and/or jongseong) will
first use a blank glyph for that jamo, and then cycle through remaining
valid code points for the respective choseong, jungseong, or jongseong.
A range can span modern and ancient, and even Hangul Jamo Extended-A
and Hangul Jamo Extended-B ranges.
.PP
For example,
.PP
.RS
.TP
-j3 11A7\-D7FB
.RE
.PP
Will first use no jongseong (because U+11A7 is one before the start of
Hangul Jamo jongseong code points), then loop through jongseong in the
Hangul Jamo range of U+11A8 through U+11FF, and then loop through
jongseong in the Hangul Jamo Extended-B range of U+D7CB through U+D7FB.
.SH OPTIONS
.RS
.IP \fBOption\fP 15
\fBFunction\fP
.TP
\fB\-h, \-\-help\fP
Print a help message and exit.
.TP
.BR \-all
Generate all Hangul syllables, using all modern and
ancient Hangul in the Unicode range U+1100..U+11FF, assigned
code points in the Extended-A range of U+A960..U+A97C, and
assigned code points in the Extended-B range of U+D7B0..U+D7FF.
\fBWARNING:\fP this will generate over 1,600,000 syllables
in a 115 megabyte Unifont .hex format file.  The default
is to only output the 11,172 modern Hangul syllables.
.TP
\fB-c\fP \fIcode_point\fP
Starting code point in hexadecimal for output file.
.TP
\fB\-j1\fP \fIstart-end\fP
Choseong (jamo 1) start-end range in hexadecimal.
.TP
\fB\-j2\fP \fIstart-end\fP
Jungseong (jamo 2) start-end range in hexadecimal.
.TP
\fB\-j3\fP \fIstart-end\fP
Jongseong (jamo 3) start-end range in hexadecimal.
.TP
\fB\-i\fP \fIinput_file\fP
Unifont hangul-base.hex formatted input file.
.TP
\fB\-o\fP \fIoutput_file\fP
Unifont .hex format output file.
.RE
.SH EXAMPLES
.TP 5
.RS
unigen-hangul -c 1 -j3 11AB-11AB \\
.PP
      -i hangul-base.hex -o nieun-only.hex
.RE
.PP
This command generates Hangul syllables using all modern choseong
and jungseong, and only the jongseong nieun (Unicode code point U+11AB).
The output Unifont .hex file will contain code points starting at 1.
Instead of specifying "-j3 11AB-11AB", simply using "-j3 11AB" will
also suffice.
.PP
This next example is a series of syllable sets suggested by Ho-Seok Ee
for preliminary syllable alignment checking of modern Hangul.
.PP
.RS
.TP 2
\(bu
The first command generates all modern syllables containing no
jongseong (final consonant), starting at Unifont hexadecimal glyph
location 0x1000; selecting a jongseong value that is out of range
(U+1160 in this case) will use a blank filler in place of the jongseong.
.TP
\(bu
The second command generates all modern syllables containing jongseong
Kiyeok (U+11AB), which has a horizontal line extending across the lower
portion of a syllable, starting at Unifont hexadecimal glyph location 0x2000.
.TP
\(bu
The third command generates all modern Hangul syllables containing
jongseong Rieul (U+11AF), starting at Unifont hexadecimal glyph location
0x3000.
.TP
\(bu
The fourth command generates all modern Hangul syllables containing
choseong (initial consonant) Rieul (U+1105), starting at Unifont hexadecimal
glyph location 0x4000.
.RE 
.PP
Here is the command sequence:
.PP
.RS
unigen-hangul -c 1000 -j1 1100-1112 -j2 1161-1175 -j3 1160 \\
.PP
      -i hangul-base.hex  >  hangul-prep.hex
.PP
unigen-hangul -c 2000 -j1 1100-1112 -j2 1161-1175 -j3 11AB \\
.PP
      -i hangul-base.hex  >> hangul-prep.hex
.PP
unigen-hangul -c 3000 -j1 1100-1112 -j2 1161-1175 -j3 11AF \\
.PP
      -i hangul-base.hex  >> hangul-prep.hex
.PP
unigen-hangul -c 4000 -j1 1105 -j2 1161-1175 -j3 11A8-11C2 \\
.PP
      -i hangul-base.hex  >> hangul-prep.hex
.RE
.PP
The resulting .hex file can then be examined with hexdraw,
unihex2bmp, \fIetc.\fP
.SH FILES
Unifont .hex files in Johab 6/3/1 encoding.  See
.BR unifont-johab631 (5)
for a description of the input file structure.  This program
uses functions contained in the file unihangul-support.c.
.SH SEE ALSO
.BR bdfimplode (1),
.BR hex2bdf (1),
.BR hex2otf (1),
.BR hex2sfd (1),
.BR hexbraille (1),
.BR hexdraw (1),
.BR hexkinya (1),
.BR hexmerge (1),
.BR johab2syllables (1),
.BR johab2ucs2 (1),
.BR unibdf2hex (1),
.BR unibmp2hex (1),
.BR unibmpbump (1),
.BR unicoverage (1),
.BR unidup (1),
.BR unifont (5),
.BR unifont-johab631 (5),
.BR unifont-viewer (1),
.BR unifont1per (1),
.BR unifontchojung (1),
.BR unifontksx (1),
.BR unifontpic (1),
.BR unigencircles (1),
.BR unigenwidth (1),
.BR unihex2bmp (1),
.BR unihex2png (1),
.BR unihexfill (1),
.BR unihexgen (1),
.BR unihexpose (1),
.BR unihexrotate (1),
.BR unijohab2html (1),
.BR unipagecount (1),
.BR unipng2hex (1)
.SH AUTHOR
.B unigen-hangul
was written by Paul Hardy.
.SH LICENSE
.B unigen-hangul
is Copyright \(co 2023 Paul Hardy.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.SH BUGS
No known bugs exist.


NEWS
2024-02-24: Release 15.1.05<F2>
  * Added almost 600 CJK ideographs, including:
    - The entire CJK Unified Ideographs Extension D range of 222 glyphs
    - 335 commonly used Cantonese ideographs in Plane 2 and Plane 3.
    - Several updates to CJK Unified Ideographs Extension I.
  * All Hangul glyphs now have a consistent look with the Hangul Syllables,
    with new glyyphs for:
    - U+1100..U+11FF Hangul Jamo
    - U+3130..U+318F Hangul Compatibility Jamo
    - U+A960..U+A97F Hangul Jamo Extended-A
    - U+D7B0..U+D7FF Hangul Jamo Extended-B.
  * Several other glyph modifications; see the ChangeLog file for details.

2023-10-29: Release 15.1.04
  * Added larger Plane 2 supersets to default and Japanese font files.

2023-10-21: Release 15.1.03
  * Boris Zhang and Yzy32767 contributed CJK Unified Ideographs
    Extension I glyphs (U+2EBF0..U+2EE5D).
  * 湖 远星 contributed 14 glyph updates to CJK Unified Ideographs
    Extension B and Extension C: U+2A6D7..U+2A6DF and
    U+2B735..U+2B739.

2023-09-21: Release 15.1.02
   * 湖 远星 adjusted 46 glyphs in the Wen Quan Yi range, U+2F00..U+9FFF, and
     contributed CJK Unified Ideographs Extension G glyphs U+30000..U+3017F.
   * Added new man pages to MAN1PAGES & MAN5PAGES definitions for installation.

2023-09-12: Release 15.1.01
   * Glyphs updated for Unicode 15.1.0 release.
   * OpenType fonts are now built by default, replacing TrueType fonts.
     TrueType fonts can be built manually by invoking "make truetype"
     in the font directory.
   * New Hangul Johab 6/3/1 encoding devised by Ho-Seok Ee for building
     the Hangul Syllables block (U+AC00..U+D7A3), with new supporting
     programs johab2syllables, unigen-hangul, and unijohab2html.  This
     encoding replaces Unifont's original Hanterm-derived encoding.
     New source files hangul.h and unihangul-support.c provide support
     for rendering syllable bitmaps with all Unicode Hangul letters,
     including all ancient Hangul, through the new unigen-hangul program.
     These functions also can support user-written Hangul syllable rendering
     programs, for example on embedded systems using graphics displays.
   * Added three hexadecimal digit representation CSUR/UCSUR scripts:
     - U+EBE0..U+EBEF: Boby Lapointe's "bibi-binary" notation
     - U+EBF0..U+EBFF: Bruce Alan Martin's bit location notation
     - U+ECF0..U+ECFF: Ronald O. Whitaker's triangular notation.
   * Modified hexdraw to support bitmap manipulation of up to 64 rows
     by 128 columns for 64-by-128 graphics display screen drawing support.
     Added the unihexpose program to transpose Unifont glyphs, for use
     with graphics display controller chips that encode each byte as
     8 rows by 1 column, left to right.
   * Other minor updates; see ChangeLog for details.

2023-06-04: Release 15.0.06
   * Further adjustments to Hangul glyphs in the Basic Multilingual Plane.

2023-06-03: Release 15.0.05
   * New Hangul Syllables block (U+AC00..U+D7A3) contributed
     by Minseo Lee.  Generated with his modifications of three
     combining patterns in the johab2ucs2 Perl script.
   * Modified U+5829 and U+6FF9 Chinese glyphs.
   * Full details in ChangeLog file.

2023-05-28: Release 15.0.04
   * Adjustments to Greek and Coptic letters.
   * Adjustments to two heart glyphs.
   * Added Engsvanáli (U+E100..U+E14F) CSUR glyphs.

2023-05-21: Release 15.0.03
   * Fixed a typo in src/Makefile that could affect
     "make install".
   * Install unifont_all.hex in $(PKGDEST) directory.
   * Other minor changes; see ChangeLog file for details.

2023-05-20: Release 15.0.02
   * N.B.: This is the last planned version to build TrueType
     font files as part of the standard font build with

           make BUILDFONT=1

     In the future, only OpenType font files will be part of
     the ordinary font building using the hex2otf program.
     TrueType fonts will have to be built manually using

           cd font
           make truetype

   * Minor update to allow for separate building of TrueType
     and OpenType fonts.  This is in preparation for the future
     removal of TrueType font files from default font builds,
     in preference for OpenType font files.  That change is
     planned for the release of Unifont 15.1.01, which will
     correspond to the release of the Unicode 15.1.0 standard
     scheduled for September 2023.
   * Added Doxygen annotation to C programs and corresponding
     ".h" header files.  Annotated output is generated in a
     new "doxygen" subdirectory.  Doxygen output consists of
     a browsable set of HTML pages and a hyperlinked PDF document.
   * Updates to make files to fix issues with redundant building
     under some circumstances.
   * Added 196 Plane 2 and Plane 3 Chinese ideographs drawn by
     湖 远星 to the basic Unifont font files for coverage of
     the General Standard Chinese Characters set.  Also added
     biang glyphs drawn by Ming Fan.
   * Fullwidth CJK characters that appeared to be boldface in
     the range U+FF00..U+FFEF were updated using glyphs from
     Wen Quan Yi, with some modifications.

2022-09-13: Release 15.0.01
   * Update for Unicode 15.0.0 release.
   * "make install" now installs Unifont OpenType font files in the
     same directory as TrueType fonts by default.  The intention is
     to remove TrueType fonts file from the "install" target for the
     Unicode 16.0.0 release unless there is a reason to keep them.
   * Existing Unicode scripts in Unifont maintain complete coverage.
   * Under ConScript Unicode Registry (UCSUR) scripts added:
     - U+E1D0..U+E1FF Syai
     - U+E360..U+E37F Sarkai
     - U+ED80..U+EDAF Iranic.
   * New Unicode 15.0.0 Plane 1 scripts:
     - U+10EFD..U+10EFF Arabic Extended-C
     - U+11B00..U+11B5F Devanagari Extended-A
     - U+11F00..U+11F5F Kawi
     - U+1D2C0..U+1D2DF Kaktovik Numerals
     - U+1E030..U+1E08F Cyrillic Exntended-D
     - U+1E4D0..U+1E4FF Nag Mundari.

2022-06-04: Release 14.0.04

src/unigen-hangul.c
/**
   @file unigen-hangul.c

   @brief Generate arbitrary hangul syllables.

   Input is a Unifont .hex file such as the "hangul-base.hex" file that
   is included in the Unifont package.

   The default program parameters will generate the Unicode
   Hangul Syllables range of U+AC00..U+D7A3.  The syllables
   will appear in this order:

        For each modern choseong {
           For each modern jungseong {
              Output syllable of choseong and jungseong
              For each modern jongseong {
                 Output syllable of choseong + jungseong + jongseong
              }
           }
        }

   By starting the jongseong code point at one before the first
   valid jongseong, the first inner loop iteration will add a
   blank glyph for the jongseong portion of the syllable, so
   only the current choseong and jungseong will be output first.

   @author Paul Hardy

   @copyright Copyright © 2023 Paul Hardy
*/
/*
   LICENSE:

      This program is free software: you can redistribute it and/or modify
      it under the terms of the GNU General Public License as published by
      the Free Software Foundation, either version 2 of the License, or
      (at your option) any later version.

      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
      GNU General Public License for more details.

      You should have received a copy of the GNU General Public License
      along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>
#include "hangul.h"

// #define DEBUG


struct PARAMS {
   unsigned starting_codept;      /* First output Unicode code point. */
   unsigned cho_start,  cho_end;  /* Choseong  start and end code points. */
   unsigned jung_start, jung_end; /* Jungseong start and end code points. */
   unsigned jong_start, jong_end; /* Jongseong start and end code points. */
   FILE *infp;
   FILE *outfp;
};


/**
   @brief Program entry point.
*/
int
main (int argc, char *argv[]) {

   int i;  /* loop variable */
   unsigned codept;
   unsigned max_codept;
   unsigned glyph[MAX_GLYPHS][16];
   unsigned tmp_glyph [16];  /* To build one combined glyph at a time. */
   int cho, jung, jong;      /* The 3 components in a Hangul syllable. */

   /// Default parameters for Hangul syllable generation.
   struct PARAMS params = { 0xAC00,  /* Starting output Unicode code point */
                            0x1100,  /* First modern choseong              */
                            0x1112,  /* Last modern choseong               */
                            0x1161,  /* First modern jungseong             */
                            0x1175,  /* Last modern jungseong              */
                            0x11A7,  /* One before first modern jongseong  */
                            0x11C2,  /* Last modern jongseong              */
                            stdin,   /* Default input file pointer         */
                            stdout   /* Default output file pointer        */
                          };

   void parse_args (int argc, char *argv[], struct PARAMS *params);

   unsigned hangul_read_base16 (FILE *infp, unsigned glyph[][16]);

   void print_glyph_hex (FILE *fp, unsigned codept, unsigned *this_glyph);

   void combined_jamo (unsigned glyph [MAX_GLYPHS][16],
                       unsigned cho, unsigned jung, unsigned jong,
                       unsigned *combined_glyph);


   if (argc > 1) {
      parse_args (argc, argv, &params);

#ifdef DEBUG
      fprintf (stderr,
               "Range: (U+%04X, U+%04X, U+%04X) to (U+%04X, U+%04X, U+%04X)\n",
               params.cho_start, params.jung_start, params.jong_start,
               params.cho_end,   params.jung_end,   params.jong_end);
#endif
   }

   /*
      Initialize glyph array to all zeroes.
   */
   for (codept = 0; codept < MAX_GLYPHS; codept++) {
      for (i = 0; i < 16; i++) glyph[codept][i] = 0x0000;
   }

   /*
      Read Hangul base glyph file.
   */
   max_codept = hangul_read_base16 (params.infp, glyph);
   if (max_codept > 0x8FF) {
      fprintf (stderr, "\nWARNING: Hangul glyph range exceeds PUA space.\n\n");
   }

   codept = params.starting_codept;  /* First code point to output */

   for (cho = params.cho_start; cho <= params.cho_end; cho++) {
      for (jung = params.jung_start; jung <= params.jung_end; jung++) {
         for (jong = params.jong_start; jong <= params.jong_end; jong++) {

#ifdef DEBUG
            fprintf (params.outfp,
                     "(U+%04X, U+%04X, U+%04X)\n",
                     cho, jung, jong);
#endif
            combined_jamo (glyph, cho, jung, jong, tmp_glyph);
            print_glyph_hex (params.outfp, codept, tmp_glyph);
            codept++;
            if (jong == JONG_UNICODE_END)
               jong = JONG_EXTB_UNICODE_START - 1; /* Start Extended-B range */
         }
         if (jung == JUNG_UNICODE_END)
            jung = JUNG_EXTB_UNICODE_START - 1;  /* Start Extended-B range */
      }
      if (cho == CHO_UNICODE_END)
         cho = CHO_EXTA_UNICODE_START - 1;  /* Start Extended-A range */
   }

   if (params.infp  != stdin)  fclose (params.infp);
   if (params.outfp != stdout) fclose (params.outfp);

   exit (EXIT_SUCCESS);
}


/**
   @brief Parse command line arguments.

*/
void
parse_args (int argc, char *argv[], struct PARAMS *params) {
   int arg_count;  /* Current index into argv[]. */

   void get_hex_range (char *instring, unsigned *start, unsigned *end);

   int strncmp (const char *s1, const char *s2, size_t n);


   arg_count = 1;

   while (arg_count < argc) {
      /* If all 600,000+ Hangul syllables are requested. */
      if (strncmp (argv [arg_count], "-all", 4) == 0) {
         params->starting_codept = 0x0001;
         params->cho_start  = CHO_UNICODE_START;       /*            First modern choseong  */
         params->cho_end    = CHO_EXTA_UNICODE_END;    /*            Last ancient choseong  */
         params->jung_start = JUNG_UNICODE_START;      /*            First modern jungseong */
         params->jung_end   = JUNG_EXTB_UNICODE_END;   /*            Last ancient jungseong */
         params->jong_start = JONG_UNICODE_START - 1;  /* One before first modern jongseong */
         params->jong_end   = JONG_EXTB_UNICODE_END;   /*            Last andient jongseong */
      }
      /* If starting code point for output Unifont hex file is specified. */
      else if (strncmp (argv [arg_count], "-c", 2) == 0) {
         arg_count++;
         if (arg_count < argc) {
            sscanf (argv [arg_count], "%X", &params->starting_codept);
         }
      }
      /* If initial consonant (choseong) range, "jamo 1", get range.  */
      else if (strncmp (argv [arg_count], "-j1", 3) == 0) {
         arg_count++;
         if (arg_count < argc) {
            get_hex_range (argv [arg_count],
                           &params->cho_start, &params->cho_end);
            /*
               Allow one initial blank glyph at start of a loop, none at end.
            */
            if (params->cho_start < CHO_UNICODE_START) {
               params->cho_start = CHO_UNICODE_START - 1;
            }
            else if (params->cho_start > CHO_UNICODE_END &&
                     params->cho_start < CHO_EXTA_UNICODE_START) {
               params->cho_start = CHO_EXTA_UNICODE_START - 1;
            }
            /*
               Do not go past desired Hangul choseong range,
               Hangul Jamo or Hangul Jamo Extended-A choseong.
            */
            if (params->cho_end > CHO_EXTA_UNICODE_END) {
               params->cho_end = CHO_EXTA_UNICODE_END;
            }
            else if (params->cho_end > CHO_UNICODE_END &&
                     params->cho_end < CHO_EXTA_UNICODE_START) {
               params->cho_end = CHO_UNICODE_END;
            }
         }
      }
      /* If medial vowel (jungseong) range, "jamo 2", get range.  */
      else if (strncmp (argv [arg_count], "-j2", 3) == 0) {
         arg_count++;
         if (arg_count < argc) {
            get_hex_range (argv [arg_count],
                           &params->jung_start, &params->jung_end);
            /*
               Allow one initial blank glyph at start of a loop, none at end.
            */
            if (params->jung_start < JUNG_UNICODE_START) {
               params->jung_start = JUNG_UNICODE_START - 1; 
            }
            else if (params->jung_start > JUNG_UNICODE_END &&
                     params->jung_start < JUNG_EXTB_UNICODE_START) {
               params->jung_start = JUNG_EXTB_UNICODE_START - 1;
            }
            /*
               Do not go past desired Hangul jungseong range,
               Hangul Jamo or Hangul Jamo Extended-B jungseong.
            */
            if (params->jung_end > JUNG_EXTB_UNICODE_END) {
               params->jung_end = JUNG_EXTB_UNICODE_END;
            }
            else if (params->jung_end > JUNG_UNICODE_END &&
                     params->jung_end < JUNG_EXTB_UNICODE_START) {
               params->jung_end = JUNG_UNICODE_END;
            }
         }
      }
      /* If final consonant (jongseong) range, "jamo 3", get range.  */
      else if (strncmp (argv [arg_count], "-j3", 3) == 0) {
         arg_count++;
         if (arg_count < argc) {
            get_hex_range (argv [arg_count],
                           &params->jong_start, &params->jong_end);
            /*
               Allow one initial blank glyph at start of a loop, none at end.
            */
            if (params->jong_start < JONG_UNICODE_START) {
               params->jong_start = JONG_UNICODE_START - 1; 
            }
            else if (params->jong_start > JONG_UNICODE_END &&
                     params->jong_start < JONG_EXTB_UNICODE_START) {
               params->jong_start = JONG_EXTB_UNICODE_START - 1;
            }
            /*
               Do not go past desired Hangul jongseong range,
               Hangul Jamo or Hangul Jamo Extended-B jongseong.
            */
            if (params->jong_end > JONG_EXTB_UNICODE_END) {
               params->jong_end = JONG_EXTB_UNICODE_END;
            }
            else if (params->jong_end > JONG_UNICODE_END &&
                     params->jong_end < JONG_EXTB_UNICODE_START) {
               params->jong_end = JONG_UNICODE_END;
            }
         }
      }
      /* If input file is specified, open it for read access. */
      else if (strncmp (argv [arg_count], "-i", 2) == 0) {
         arg_count++;
         if (arg_count < argc) {
            params->infp = fopen (argv [arg_count], "r");
            if (params->infp == NULL) {
               fprintf (stderr, "\n*** ERROR: Cannot open %s for input.\n\n",
                        argv [arg_count]);
               exit (EXIT_FAILURE);
            }
         }
      }
      /* If output file is specified, open it for write access. */
      else if (strncmp (argv [arg_count], "-o", 2) == 0) {
         arg_count++;
         if (arg_count < argc) {
            params->outfp = fopen (argv [arg_count], "w");
            if (params->outfp == NULL) {
               fprintf (stderr, "\n*** ERROR: Cannot open %s for output.\n\n",
                        argv [arg_count]);
               exit (EXIT_FAILURE);
            }
         }
      }
      /* If help is requested, print help message and exit. */
      else if (strncmp (argv [arg_count], "-h",     2) == 0 ||
               strncmp (argv [arg_count], "--help", 6) == 0) {
         printf ("\nunigen-hangul [options]\n\n");
         printf ("     Generates Hangul syllables from an input Unifont .hex file encoded\n");
         printf ("     in Johab 6/3/1 format.  By default, the output is the Unicode Hangul\n");
         printf ("     Syllables range, U+AC00..U+D7A3.  Options allow the user to specify\n");
         printf ("     a starting code point for the output Unifont .hex file, and ranges\n");
         printf ("     in hexadecimal of the starting and ending Hangul Jamo code points:\n\n");

         printf ("          * 1100-115E Initial consonants (choseong)\n");
         printf ("          * 1161-11A7 Medial vowels (jungseong)\n");
         printf ("          * 11A8-11FF Final consonants (jongseong).\n\n");

         printf ("     A single code point or 0 to omit can be specified instead of a range.\n\n");

         printf ("   Option    Parameters    Function\n");
         printf ("   ------    ----------    --------\n");
         printf ("   -h, --help              Print this message and exit.\n\n");
         printf ("   -all                    Generate all Hangul syllables, using all modern and\n");
         printf ("                           ancient Hangul in the Unicode range U+1100..U+11FF,\n");
         printf ("                           U+A960..U+A97C, and U+D7B0..U+D7FB.\n");
         printf ("                           WARNING: this will generate over 1,600,000 syllables\n");
         printf ("                           in a 115 megabyte Unifont .hex format file.  The\n");
         printf ("                           default is to only output modern Hangul syllables.\n\n");
         printf ("   -c        code_point    Starting code point in hexadecimal for output file.\n\n");
         printf ("   -j1       start-end     Choseong (jamo 1) start-end range in hexadecimal.\n\n");
         printf ("   -j2       start-end     Jungseong (jamo 2) start-end range in hexadecimal.\n\n");
         printf ("   -j3       start-end     Jongseong (jamo 3) start-end range in hexadecimal.\n\n");
         printf ("   -i        input_file    Unifont hangul-base.hex formatted input file.\n\n");
         printf ("   -o        output_file   Unifont .hex format output file.\n\n");
         printf ("      Example:\n\n");
         printf ("         unigen-hangul -c 1 -j3 11AB-11AB -i hangul-base.hex -o nieun-only.hex\n\n");
         printf ("      Generates Hangul syllables using all modern choseong and jungseong,\n");
         printf ("      and only the jongseong nieun (Unicode code point U+11AB).  The output\n");
         printf ("      Unifont .hex file will contain code points starting at 1.  Instead of\n");
         printf ("      specifying \"-j3 11AB-11AB\", simply using \"-j3 11AB\" will also suffice.\n\n");

         exit (EXIT_SUCCESS);
      }

      arg_count++;
   }

   return;
}


/**
   @brief Scan a hexadecimal range from a character string.
*/
void
get_hex_range (char *instring, unsigned *start, unsigned *end) {

   int i;  /* String index variable. */

   /* Get first number in range. */
   sscanf (instring, "%X", start);
   for (i = 0;
        instring [i] != '\0' && instring [i] != '-';
        i++);
   /* Get last number in range. */
   if (instring [i] == '-') {
      i++;
      sscanf (&instring [i], "%X", end);
   }
   else {
      *end = *start;
   }

   return;
}
